src/compose.ts

```diff
@@ -12,33 +14,41 @@ import type { Last } from './types/last'
  * matches the type of the previous function's argument.
  */
 export type Composition<
-  Functions extends Array<(...args: any[]) => any>,
+  Functions extends Array<Fn>,
   Length extends number = Functions['length']
   > =
   Length extends 1
-    ? Functions
-    : Functions extends [...infer Rest, infer Penultimate, infer Last]
-      ? [
-        ...Composition<
-          // Weirdly `Rest`, `Penultimate` and `Last`,
-          // which we extended from unary function type,
-          // are not considered as functions in this expression
-          // @ts-ignore
-          [...Rest, (arg: ReturnType<Last>) => ReturnType<Penultimate>]
-          >,
-        Last
-      ]
-      : any
+  ? Functions
+  : Functions extends [...infer Rest, infer Penultimate, infer Last]
+  ? [
+    ...Composition<
+      // Weirdly `Rest`, `Penultimate` and `Last`,
+      // which we extended from unary function type,
+      // are not considered as functions in this expression
+      Rest extends Array<Fn>
+      ? Last extends Fn
+      ? Penultimate extends Fn
+      ? [...Rest, (arg: ReturnType<Last>) => ReturnType<Penultimate>]
+      : never
+      : never
+      : never
+    >,
+    Last
+  ]
+  : Functions
 
 /**
  * Applies all of the provided `functions` one-by-one in right-to-left order
  * starting from the `argument`.
  */
-export default function compose<Functions extends Array<(arg: any) => any>> (
+export default function compose<Functions extends Array<Fn>>(
```

Comment:
`Fn` type here is actually incorrect, as `compose` directly implies that we're working with **unary** functions (accept exactly **one** argument) here, and the `Fn` is a variadic one (accepts an arbitrary amount of arguments).